typescript/sdk/src/warp/WarpCore.ts

```diff
@@ -0,0 +1,419 @@
+    });
+  }
+
+  async getTransferGasQuote(
+    originToken: Token,
+    destination: ChainNameOrId,
+  ): Promise<TokenAmount> {
+    const { chainName: originName, protocol: originProtocol } = originToken;
+    const destinationName = this.multiProvider.getChainName(destination);
+
+    // Step 1: Determine the amount
+
+    let gasAmount: bigint;
+    // Check constant quotes first
+    const defaultQuote = this.igpQuoteConstants.find(
+      (q) => q.origin === originName && q.destination === destinationName,
+    );
+    if (defaultQuote) {
+      gasAmount = BigInt(defaultQuote.quote.toString());
+    } else {
+      // Otherwise, compute IGP quote via the adapter
+      const hypAdapter = originToken.getHypAdapter(this.multiProvider);
+      const destinationDomainId = this.multiProvider.getDomainId(destination);
+      gasAmount = BigInt(await hypAdapter.quoteGasPayment(destinationDomainId));
+    }
+
+    // Step 2: Determine the IGP token
+    // TODO, it would be more robust to determine this based on on-chain data
+    // rather than these janky heuristic
+
+    let igpToken: Token;
+    if (
+      originToken.igpTokenAddressOrDenom ||
+      (originToken.collateralAddressOrDenom &&
+        originProtocol === ProtocolType.Cosmos)
+    ) {
+      const address =
+        originToken.igpTokenAddressOrDenom ||
+        originToken.collateralAddressOrDenom;
+      const searchResult = this.findToken(originName, address);
+      if (!searchResult) throw new Error(`IGP token ${address} is unknown`);
+      igpToken = searchResult;
+    } else {
+      // Otherwise use the plain old native token from the route origin
+      igpToken = Token.FromChainMetadataNativeToken(
+        this.multiProvider.getChainMetadata(originName),
+      );
+    }
+
+    this.logger(`Quoted igp gas payment: ${gasAmount} ${igpToken.symbol}`);
+    return new TokenAmount(gasAmount, igpToken);
+  }
+
+  async getTransferRemoteTxs(
+    originTokenAmount: TokenAmount,
+    destination: ChainNameOrId,
+    sender: Address,
+    recipient: Address,
+  ): Promise<{ approveTx?: TypedTransaction; transferTx: TypedTransaction }> {
+    const { token, amount } = originTokenAmount;
+    const destinationDomainId = this.multiProvider.getDomainId(destination);
+    const providerType = PROTOCOL_TO_DEFAULT_PROVIDER_TYPE[token.protocol];
+    const hypAdapter = token.getHypAdapter(this.multiProvider);
+
+    let approveTx: TypedTransaction | undefined = undefined;
+    if (await this.isApproveRequired(originTokenAmount, sender)) {
+      this.logger(`Approval required for transfer of ${token.symbol}`);
+      const approveTxReq = await hypAdapter.populateApproveTx({
+        weiAmountOrId: amount.toString(),
+        recipient: token.addressOrDenom,
+      });
+      this.logger(`Approval tx for ${token.symbol} populated`);
+      approveTx = {
+        type: providerType,
+        transaction: approveTxReq,
+      } as TypedTransaction;
+    }
```

Comment:
maybe move this into `hypTokenAdapter.populateTransferRemote`

Reply:
Keeping this here for now but the logic to check for approvals was indeed moved into the adapters.